rust/ql/lib/codeql/rust/elements/internal/AstNodeImpl.qll

```diff
@@ -12,5 +11,36 @@ private import codeql.rust.elements.internal.generated.AstNode
  * be referenced directly.
  */
 module Impl {
-  class AstNode extends Generated::AstNode { }
+  private import rust
+  private import codeql.rust.elements.internal.generated.ParentChild
+
+  /**
+   * Gets the immediate parent of a non-`AstNode` element `e`.
+   *
+   * We restrict `e` to be a non-`AstNode` to skip past non-`AstNode` in
+   * the transitive closure computation in `getParentOfAst`. This is
+   * necessary because the parent of an `AstNode` is not necessarily an `AstNode`.
+   */
+  private Element getParentOfAstStep(Element e) {
+    not e instanceof AstNode and
+    result = getImmediateParent(e)
+  }
+
+  /** Gets the nearest enclosing parent of `ast` that is an `AstNode`. */
+  private AstNode getParentOfAst(AstNode ast) {
+    result = getParentOfAstStep*(getImmediateParent(ast))
+  }
+
+  class AstNode extends Generated::AstNode {
+    /** Gets the immediately enclosing callable of this node, if any. */
+    cached
+    Callable getEnclosingCallable() {
```

Comment:
Shall we also add `AstNode getParent()` ?

Reply:
Good idea.